lib/passport-azure-ad/metadata.js

```diff
@@ -102,41 +121,128 @@ Metadata.prototype.updateWsfedMetadata = function(doc, next) {
+      },
+
+      function(body, next) {
+        // parse the AAD Federation metadata xml
+        log.info("Parsing JSON retreived from the signing keys endpoint.");
+        try {
+          self.oidc.keys = JSON.parse(body).keys;
+          log.info("***** KEYS  ******");
+          log.info("");
+          log.info(self.oidc.keys);
+          log.info("");
+          log.info("***********");
+          next(null);
+        } catch (e) {
+          next(new Error(e));
+        }
+      },
+
+    ], function(err) {
+      callback(err);
+    });
+
+  } catch (e) {
+    next(new Error('Invalid Open ID Connect Federation Metadata ' + e.message));
+  }
+};
+
+Metadata.prototype.generateOidcPEM = function(kid) {
+  if (!this.oidc.keys) {
+    return null;
+  }
+  for (var i = 0; i < this.oidc.keys.length; i++) {
+    if (this.oidc.keys[i].kid === kid) {
+      log.info('Working on key: ', this.oidc.keys[i]);
+      var modulus = new Buffer(this.oidc.keys[i].n, 'base64');
+      var exponent = new Buffer(this.oidc.keys[i].e, 'base64');
+
+      var pubKey = ursa.createPublicKeyFromComponents(modulus, exponent);
+      return pubKey.toPublicPem('utf8');
+    }
+  }
+  return null;
+};
+
 Metadata.prototype.fetch = function(callback) {
   var self = this;
 
   async.waterfall([
     // fetch the Federation metadata for the AAD tenant
-    function(next){
-      request(self.url, function (err, response, body) {
-        if(err) {
+    function(next) {
+      request(self.url, function(err, response, body) {
+        if (err) {
           next(err);
-        } else if(response.statusCode !== 200) {
-          next(new Error("Error:" + response.statusCode +  " Cannot get AAD Federation metadata from " + self.url));
+        } else if (response.statusCode !== 200) {
+          next(new Error("Error:" + response.statusCode + " Cannot get AAD Federation metadata from " + self.url));
         } else {
           next(null, body);
         }
       });
     },
-    function(body, next){
-      // parse the AAD Federation metadata xml
-      var parser = new xml2js.Parser({explicitRoot:true});
-      // Note: xml responses from Azure AAD have a leading \ufeff which breaks xml2js parser!
-      parser.parseString(body.replace("\ufeff", ""), function (err, data) {
-        self.metatdata = data;
-        next(err);
+    function(body, next) {
+      if (self.authtype === "saml" || self.authtype === "wsfed") {
+        // parse the AAD Federation metadata xml
+        var parser = new xml2js.Parser({
+          explicitRoot: true
+        });
+        // Note: xml responses from Azure AAD have a leading \ufeff which breaks xml2js parser!
+        parser.parseString(body.replace("\ufeff", ""), function(err, data) {
+          self.metatdata = data;
+          next(err);
+
+        });
+      } else if (self.authtype === "oidc") {
+        log.info("Parsing JSON retreived from the endpoint");
+        self.metadata = JSON.parse(body);
+        next(null);
```

Comment:
SAML and WSFED uses XML and so the XMLParser has an err function that is returned we need to handle. The JSON parser is built in to node.js and does not release an err function.